MasterApeAdmin

DeFiFoFum opened the pull request:
The `MasterApeAdmin` contract is an admin helper proxy contract to the MasterApe which adds the following functionality: 

- Add batch MasterApe farms
- Set batch MasterApe farms
- Add "Fixed Percentage Farms": These are farms that need to take a constant percentage of the newly minted BANANA supply. One use is to support "ghost farms" without needing to constantly update allocations. 
- Update/Remove "Fixed Percentage Farms": Change percentage allocations 

When new farms are added or updated through the `MasterApeAdmin` contract, the fixed percentage allocation farms are updated directly after to keep their percentages constant.

DeFiFoFum committed:
feat: Add initial MasterApe admin contract

DeFiFoFum committed:
refactor: Order layout and improve logic
- Add BASE_PERCENTAGE variable to keep track of BANANA pool allocation
- Update logic for ghost farm allocation updates
- Update interface
- Add style guide

DeFiFoFum committed:
feat: Add IBananaAllocator

DeFiFoFum committed:
refactor: Use OZ contracts

DeFiFoFum committed:
test: MasterApeAdmin
- Add @apeswapfinance/test-helpers
- Add initial tests for MasterApeAdmin
- Add .npmrc to pull package. Export your read github tokens at NPM_TOKEN

DeFiFoFum committed:
refactor: Naming, cleaning, testing, improving
- Rename "GhostFarm" to "FixedPercentageFarm" to be more gneric
- Cleaned and reorded events
- PERCENTAGE_PRECISION has been fixed to 1e3
- Add state variables to assit logic
- Clean code

DeFiFoFum committed:
tests: Add tests to MasterApeAdmin

DeFiFoFum committed:
test: Add MasterApeAdmin tests
- Provide sweep token function in MasterApeAdmin
- onlyFarmAdmin tests
- onlyOwner test
- test adding/updating and removing fixed percentage farms

DeFiFoFum committed:
feat: natspec + helper functions
- Add netspec comments to each function/state variable
- Rename internal functions to match MasterApe style
- Add getDetailePoolInfo helper function to see more detailed allocation info for each pid
- Add batchUpdateMasterApePools to be able to pass an array of pids and update multiple pools
   This is to avoid a future scenario where massUpdatePools reaches the block gas limit
- Update PERCENTAGE_PRECISION to 1e4 to be able to set hundreths percetanges on fixed percentage farms

DeFiFoFum committed:
refactor: MasterApeAdmin
- Onwership is now transferred through a pending/accept flow
- Removed setting MasterApe pid allocation to 0 when disabling a fixed percentage farm
- Update tests

DeFiFoFum committed:
refactor: Update return structure of detailed pool info

DeFiFoFum committed:
feat: Check that fixed farms are not for pid 0 (reserved)

DeFiFoFum committed:
refactor: rename to sweepTokens

DeFiFoFum committed:
feat: Fixed farm syncing is now optional on updates

DeFiFoFum committed:
refactor: Script name

DeFiFoFum committed:
feat: Add admin migration script

Apegurus commented on contracts/MasterApeAdmin.sol:
```diff
@@ -0,0 +1,356 @@
+    /// @return poolBananaPerBlock Amount of BANANA given to the pool per block
+    /// @return poolBananaPerDay Amount of BANANA given to the pool per day
+    /// @return poolBananaPerMonth Amount of BANANA given to the pool per month
+    function getDetailedPoolInfo(uint pid) external view returns (
```
Very nice helper function!

Apegurus commented on contracts/MasterApeAdmin.sol:
```diff
@@ -0,0 +1,356 @@
+
+    /// @notice Run through fixed percentage farm allocations and set MasterApe allocations to match the percentage.
+    /// @dev The MasterApe contract manages the BANANA pool percentage on its own which is accounted for in the calculations below.
+    function _syncFixedPercentFarms() internal {
```
Should we add the possibility to send a range to this function? 

Thus not bricking the contract if a gas limit is hit.

DeFiFoFum replied:
We can find ways to improve the gas efficiency, but the current calculation depends on updating them all at once.

Also, if this function doesn't work, it won't brick the contract as you can still make farm updates without syncing and we can remove fixed percentage farms if there are too many.

Apegurus replied:
Got it!

Apegurus reviewed (changes requested):
Would be nice to add the possibility to send a range/array of PIDs to avoid the contract being bricked by gas limits

DeFiFoFum committed:
docs: Update comments

DeFiFoFum committed:
docs: Add preliminary audit report MasterApeAdmin

DeFiFoFum committed:
refactor: Add max bonus multiplier check

DeFiFoFum committed:
refactor: Update masterApe to immutable

DeFiFoFum committed:
refactor: Audit review
- Add require statement to external sync function
- Add local variables where needed to remove multiple contract reads

DeFiFoFum committed:
docs: Add revised audit report

DeFiFoFum committed:
Merge pull request #25 from ApeSwapFinance/audit/masterape-admin

MasterApeAdmin Audit Review
